File: frontend/utils/routes/transaction.ts
import type { AxiosResponse } from 'axios';
import { StringResponseWithStatus } from '../responses';
import axios from 'axios';
import { url } from '../constants';

export type Split = {
    tx_id: string,
    user: string,
    share: string,
}

export type RequestSplit = {
    address: string,
    share: string,
}

export type RequestTransaction = {
    name: string,
    group: string,
    total: string,
    currency: string,
    splits: Array<RequestSplit> 
}

export type Transaction = {
    name: string,
    group: string,
    total: string,
    currency: string,
    image?: string,
    splits: Array<Split>
}

export type DbTransaction = {
    id: string,
    group: string,
    amount: string,
    currency: string,
    paid_by: string,
    name: string,
    date: string,
    updated_at: string,
}

export type TransactionWithSplits = {
    transaction: DbTransaction,
    splits: Array<Split>
}

export const createTransaction = (transaction: RequestTransaction): Promise<AxiosResponse<StringResponseWithStatus>> => {
    return axios.post(`${url}/transaction`, transaction);
}

export const updateTransaction = (transaction: Transaction, tx_id: string): Promise<AxiosResponse<StringResponseWithStatus>> => {
    return axios.put(`${url}/transaction/${tx_id}`, transaction);
}

export const getTransactionsByGroup = (groupId: string): Promise<AxiosResponse<Array<DbTransaction>>> => {
    return axios.get(`${url}/transaction/group/${groupId}`);
}

export const getTransactionsByGroupWithSplits = (groupId: string): Promise<AxiosResponse<Array<TransactionWithSplits>>> => {
    return axios.get(`${url}/transaction/withsplits/group/${groupId}`);
}

File: backend/src/models/transaction.rs
use chrono::NaiveDate;
use rust_decimal::Decimal;

use crate::models::split::Split;
use serde::{Deserialize, Serialize};

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct DbTransaction {
    pub id: String,
    pub group: String,
    pub base_amount: Decimal,
    pub amount: Decimal,
    pub currency: String,
    pub paid_by: String,
    pub name: String,
    pub date: NaiveDate,
    pub updated_at: chrono::DateTime<chrono::Utc>,
}

pub struct DbTransactionWithSplits {
    pub id: String,
    pub group: String,
    pub base_amount: Decimal,
    pub amount: Decimal,
    pub currency: String,
    pub paid_by: String,
    pub name: String,
    pub date: NaiveDate,
    pub updated_at: chrono::DateTime<chrono::Utc>,
    pub tx_id: String,
    pub user: String,
    pub base_share: Decimal,
    pub share: Decimal,
    pub resolved: i8,
}

#[derive(Debug, Deserialize, Serialize, Clone)]
pub struct TransactionWithSplits {
    pub transaction: DbTransaction,
    pub splits: Vec<Split>,
}
